README.md
# 烟火活动的红包雨

## 引入

```shell
$ npm i @tencent/firework-redpack-rain
```

## 样例

您可以访问[http://firework-redpack-rain.pages.oa.com](http://firework-redpack-rain.pages.oa.com)，查看样例。

## 使用

简要使用：

```javascript
import RedpackRain from '@tencent/firework-redpack-rain';

const rain = new RedpackRain({
  selector: '.rain-container', // 要嵌入的容器
  onClick(isHited) {
    // 点击回调，只要点击就会回调，isHited为boolean类型，表示是否击中红包
    console.log(isHited);
  },
});
rain.start(); // 红包雨开始

setTimeout(function () {
  rain.stop(); // 红包雨结束
}, 6000);
```

在 react 中使用：

```javascript
function App() {
  const rainRef = (useRef < null) | (RedpackRain > null);
  useEffect(() => {
    const rain = new RedpackRain({
      selector: '.rain-container',
    });
    rainRef.current = rain;
    rain.start();

    return () => rain.stop();
  }, []);

  const handleStopClick = () => {
    if (rainRef.current) {
      rainRef.current.stop();
    }
  };

  return (
    <div className="App">
      <button onClick={handleStopClick}>停止</button>
      <div className="rain-container"></div>
    </div>
  );
}
```

## 实例方法说明

实例有 3 个方法。

### start

初始化实例后，并不会立刻启动红包雨，而是执行`start()`后，才开始下红包雨。

### stop

调用`stop()`方法后，即停止下红包雨，并清空 canvas 面板。

### setOptions

可以在下红包雨的过程中，修改配置，需要注意的是：

1. 该方法需要在 start()之后和 stop()之前才能调用，即在下红包雨的过程中才有效；
2. 新的配置只会对后续新产生的的红包有影响，之前已产生红包依然使用之前的配置；

输入的参数与初始化时的参数几乎一样（除了无法设置 selector 外，其他的参数均可以重新设置）；而且，二级配置中也可以只设置部分参数；如下面的配置中，redpack 配置中有多个参数，我们可以只设置 speedMin 和 speedMax 两个参数即可。

```javascript
rain.setOptions({
  interval: 800, // 修改红包产生的时间间隔，单位毫秒
  redpack: {
    speedMin: 40, // 红包下降速度的下限
    speedMax: 40, // 红包下降速度的上限
  },
});
```

其他参数，可以参照下面的`参数说明`。

## 参数说明

`RedpackRain`是一个类，在构建实例时，除`selector`外，其他均为非必须参数。

| 参数                | 是否必填 | 类型                         | 默认值 | 说明                                                 |
| ------------------- | -------- | ---------------------------- | ------ | ---------------------------------------------------- |
| selector            | 是       | string \| HTMLElement        |        | 要渲染红包雨的容器<br/>可以是选择器也可以是 dom 元素 |
| interval            | 否       | number                       | 1600   | 下红包雨的间隔，单位毫秒                             |
| onClick             |          | function(isHited: boolean){} | 空     | 整个红包雨区域的点击，isHited 表示是否命中红包       |
| onMonitor           |          | function({fps}){}            | 空     | 红包雨的 fps 监控                                    |
| redpack             | 否       |                              |        | 红包配置                                             |
| redpack.speedMin    | 否       | number                       | 10     | 红包下降速度的最小值                                 |
| redpack.speedMax    | 否       | number                       | 10     | 红包下降速度的最大值                                 |
| redpack.imgUrl      | 否       | string                       |        | 红包的图片                                           |
| redpack.width       |          | number                       | 192    | 红包的宽度                                           |
| redpack.height      |          | number                       | 216    | 红包的高度                                           |
| bubble              |          |                              |        | 命中红包后的上升气泡的配置                           |
| bubble.imgUrl       |          | string                       |        | 气泡的图片                                           |
| bubble.width        |          | number                       | 156    | 气泡的宽度                                           |
| bubble.height       |          | number                       | 111    | 气泡的高度                                           |
| bubble.speed        |          | number                       | 5      | 气泡每帧上升的高度                                   |
| bubble.opacitySpeed |          | number                       | 0.04   | 气泡每帧减少的透明度                                 |

使用方式：

```javascript
const rain = new RedpackRain({
  selector: document.body,
  interval: 1600,
  redpack: {
    speedMin: 10,
    speedMax: 10,
    imgUrl: 'https://sola.gtimg.cn/aoi/sola/20201226100322_I1ltnkzJVc.png',
    width: 126,
    height: 174,
  },
  bubble: {
    imgUrl: 'https://sola.gtimg.cn/aoi/sola/20201225103914_2QQ9bXg2rU.png',
    width: 156,
    height: 111,
    speed: 5,
    opacitySpeed: 0.04,
  },
  onClick: (isHited) => {
    console.log(isHited);
  },
  onMonitor(monitors) {
    console.log(monitors);
  },
});
```


CHANGELOG.md
## 1.0.8

- fix: 修复调用 setOptions 修改参数时，会产生多个 visibilityChange 事件，进而导致产生多个 setInterval；

## 1.0.7

- fix: 添加一个 clear()方法，仅擦除面板，但还保留数据，便于用于修改 interval 参数；

## 1.0.6

- feat: 添加 setOptions 方法，可以在下红包雨的过程中修改后续红包的行为；
- fix: 当调用 stop()方法后应当清除之前存储的红包数据；

## 1.0.5

- fix: 添加坐标判断，提高点击的精度；
- fix: 调整红包产生的位置，只在距离屏幕边缘 redpackWidth 的中间位置产生红包；
- feat: 添加 touchstart 类型；

## 1.0.2

- feat: 将每个红包的 requestAnimationFrame 提升到主流程中；
- feat: 添加 fps 的监控；
- fix: 页面不可见时不创建新的红包；


__tests__/index.test.ts
import { sum, subtraction } from "../src/index";

describe("sum", () => {
  test("sum case 1", () => {
    expect(sum(1, 2)).toBe(3);
  });
  test("sum case 2", () => {
    expect(sum(1, -2)).toBe(-1);
  });
  test("sum case 3", () => {
    expect(sum(-1, -2)).toBe(-3);
  });
});

describe("subtraction", () => {
  test("subtraction case 1", () => {
    expect(subtraction(1, 2)).toBe(-1);
  });
  test("subtraction case 2", () => {
    expect(subtraction(1, -2)).toBe(3);
  });
  test("subtraction case 3", () => {
    expect(subtraction(-1, -2)).toBe(1);
  });
});


src/index.ts
import defaultsConfig from './config/defaults';
import RedpackItem from './libs/redpackItem';
import { requestAnimationFramePolyfill, cancelAnimationFramePolyfill } from './utils/animationFrame';
import Visibility from 'gh-qqnews-utils/visibility';

// 每个红包的配置
interface RedpackRainItem {
  speedMin?: number;
  speedMax?: number;
  imgUrl?: string;
  width?: number;
  height?: number;
}

interface BubbleProps {
  imgUrl?: string;
  width?: number;
  height?: number;
  speed?: number;
  opacitySpeed?: number;
}

interface MonitorProps {
  fps: number;
}

interface RedpackRainOptions {
  interval?: number;
  eventType?: 'click' | 'touchstart';
  redpack?: RedpackRainItem;
  bubble?: BubbleProps;
  onClick?: (isHit: boolean) => void;
  onMonitor?: (props: MonitorProps) => void;
}

interface RedpackRainProps extends RedpackRainOptions {
  selector: HTMLElement | string;
}

class RedpackRain {
  private timer: NodeJS.Timeout | null = null;
  private rainCtx: CanvasRenderingContext2D | null = null;
  private bubbleCtx: CanvasRenderingContext2D | null = null;
  private config = defaultsConfig;
  private parentClientRect = { width: 0, height: 0, top: 0, left: 0 };
  private ratio = 3;
  private redpackItemList: {
    [key: number]: RedpackItem;
  } = {};
  private requestId: number | null = null;
  private fpsBefore = Date.now(); // 计算fps
  private lastRedpackX = 0; // 上个红包的横坐标
  private pageVisibility: Visibility | null = null;

  constructor(props: RedpackRainProps) {
    this.createConfig(props);

    if (typeof props.selector === 'string') {
      const dom: HTMLElement | null = document.querySelector(props.selector);
      if (dom) {
        this.config.selector = dom;
      } else {
        console.error(`rain container cant found, selector: ${props.selector}`);
      }
    } else {
      this.config.selector = props.selector;
    }
    this.creatCanvas();
    this.pageVisibility = new Visibility();
  }

  // 手动配置覆盖掉默认配置
  private createConfig(config: any, parentKey = '') {
    // eslint-disable-next-line no-restricted-syntax
    for (const key in config) {
      const item = config[key];

      if (item) {
        if (typeof item === 'object') {
          this.createConfig(item, key);
        } else {
          if (parentKey) {
            (this.config as any)[parentKey][key] = item;
          } else {
            (this.config as any)[key] = item;
          }
        }
      }
    }
  }

  // 在指定容器内创建2个canvas
  private creatCanvas() {
    const selector: HTMLElement = this.config.selector as HTMLElement;

    const { top, left, width, height } = selector.getBoundingClientRect();

    this.parentClientRect.width = width * this.ratio;
    this.parentClientRect.height = height * this.ratio;
    this.parentClientRect.top = top;
    this.parentClientRect.left = left;

    if (selector.getElementsByTagName('canvas').length === 0) {
      const canvasBubble = document.createElement('canvas');
      canvasBubble.className = 'bubble-redpack-canvas';
      canvasBubble.style.cssText = 'position: absolute; top: 0; left: 0; width: 100%; height: 100%; z-index: 1';
      canvasBubble.width = this.parentClientRect.width;
      canvasBubble.height = this.parentClientRect.height;

      const canvasRain = document.createElement('canvas');
      canvasRain.className = 'rain-redpack-canvas';
      canvasRain.style.cssText = 'position: absolute; top: 0; left: 0; width: 100%; height: 100%; z-index: 2';
      canvasRain.width = this.parentClientRect.width;
      canvasRain.height = this.parentClientRect.height;

      const div = document.createElement('div');
      div.className = `rain-item rain-${Date.now()}-${Math.random()}`;
      div.style.cssText = 'position: relative; height: 100%';
      div.appendChild(canvasBubble);
      div.appendChild(canvasRain);

      selector.appendChild(div);

      const rainCanvasSelector: HTMLCanvasElement | null = selector.querySelector('.rain-redpack-canvas');
      const bubbleCanvasSelector: HTMLCanvasElement | null = selector.querySelector('.bubble-redpack-canvas');

      if (rainCanvasSelector && bubbleCanvasSelector) {
        this.rainCtx = rainCanvasSelector.getContext('2d');
        this.bubbleCtx = bubbleCanvasSelector.getContext('2d');
      }
    }
  }

  /**
   * 返回红包创建时的x轴坐标
   * https://git.code.oa.com/news-gh-team/firework-redpack-rain/issues/2
   * @param width 红包的宽度
   */
  private getRedpackItemX(width: number): number {
    let x = this.lastRedpackX;
    do {
      x = Math.floor(Math.random() * (this.parentClientRect.width - width * 3) + width); // 避免红包产生在边界
    } while (Math.abs(this.lastRedpackX - x) <= width * 1.5); // 避免先后两个红包重叠

    this.lastRedpackX = x;
    return x;
  }

  // 创建红包
  private createRedpackItem() {
    if (!this.rainCtx || !this.bubbleCtx) {
      this.stop();
      return;
    }
    const { width, height, speedMax, speedMin, imgUrl } = this.config.redpack || {};
    const redpackItemId = Date.now();
    const x = this.getRedpackItemX(width);
    const redpackItem = new RedpackItem({
      redpackId: redpackItemId,
      redpackCtx: this.rainCtx,
      bubbleCtx: this.bubbleCtx,
      x,
      y: -this.config.redpack.height,
      redpackImgUrl: imgUrl,
      width,
      height,
      speedMax,
      speedMin,
      bubble: this.config.bubble,
      containerHeight: this.parentClientRect.height,
      onDestoryed: (id) => {
        delete this.redpackItemList[id];
      },
    });
    redpackItem.start();
    this.redpackItemList[redpackItemId] = redpackItem;
  }

  private clickListener = (event: MouseEvent | TouchEvent) => {
    if (typeof this.config.onClick !== 'function') {
      return;
    }
    let touchClients: Array<{ clientX: number; clientY: number }> = []; // 手指触控的坐标
    if (event.type === 'touchstart') {
      // touchstart
      touchClients = Array.prototype.slice.call((event as TouchEvent).touches).map(touch => ({
        clientX: touch.clientX,
        clientY: touch.clientY,
      }));
    } else {
      // click
      touchClients = [
        {
          clientX: (event as MouseEvent).clientX,
          clientY: (event as MouseEvent).clientY,
        },
      ];
    }

    const { left, top } = this.parentClientRect;

    if (this.requestId) {
      cancelAnimationFramePolyfill(this.requestId);
    }
    touchClients.forEach(({ clientX, clientY }) => {
      const myClientX = (clientX - left) * this.ratio;
      const myClientY = (clientY - top) * this.ratio;
      // eslint-disable-next-line no-restricted-syntax
      for (const key in this.redpackItemList) {
        const redpackItem = this.redpackItemList[key];
        const { x, y, width, height } = redpackItem;

        const diff = 14;
        let isHited = false;
        if (
          myClientX >= x - diff
          && myClientX <= x + width + diff
          && myClientY >= y - diff
          && myClientY <= y + height + diff
        ) {
          delete this.redpackItemList[key];
          redpackItem.addBubble();
          isHited = true;
        }
        if (typeof this.config.onClick === 'function') {
          this.config.onClick(isHited);
        }
      }
    });
    this.render();
  };

  private render() {
    this.requestId = requestAnimationFramePolyfill(() => {
      this.rainCtx?.clearRect(0, 0, this.parentClientRect.width, this.parentClientRect.height);
      // eslint-disable-next-line no-restricted-syntax
      for (const key in this.redpackItemList) {
        const redpackItem = this.redpackItemList[key];

        redpackItem.render();
      }
      const now = Date.now();
      const fps = Math.round(1000 / (now - this.fpsBefore));
      this.fpsBefore = now;
      if (typeof this.config.onMonitor === 'function') {
        this.config.onMonitor({ fps });
      }
      this.render();
    });
  }

  // eslint-disable-next-line @typescript-eslint/member-ordering
  setOptions(options: RedpackRainOptions) {
    if (!this.timer) {
      // 若没有启动红包雨，则无法调用该方法
      return console.error('please use start() before setOptions');
    }
    const beforeInterval = this.config.interval;
    this.createConfig(options);

    // 当下红包雨的间隔变化后，则重置定时器；
    // 若间隔不变，则还是使用之前的定时器
    if (beforeInterval !== this.config.interval) {
      this.start();
    }
  }

  // eslint-disable-next-line @typescript-eslint/member-ordering
  start() {
    // 先停止上一个
    this.clear();
    this.config.selector.addEventListener(this.config.eventType, this.clickListener, false);

    this.createRedpackItem();

    // 创建一个新的红包雨
    this.timer = setInterval(() => {
      this.createRedpackItem();
    }, this.config.interval);
    this.render();

    this.pageVisibility?.visibilityChange((isShow) => {
      if (this.timer) {
        clearInterval(this.timer);
        this.timer = null;
      }
      if (isShow) {
        // 创建一个新的红包雨
        this.timer = setInterval(() => {
          this.createRedpackItem();
        }, this.config.interval);
      }
    });
  }

  // eslint-disable-next-line @typescript-eslint/member-ordering
  clear() {
    // 停止产生新的红包
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }

    // 停止渲染
    if (this.requestId) {
      cancelAnimationFramePolyfill(this.requestId);
      this.requestId = null;
    }

    // 清空面板
    this.rainCtx?.clearRect(0, 0, this.parentClientRect.width, this.parentClientRect.height);

    // 解除已绑定的事件
    this.config.selector.removeEventListener(this.config.eventType, this.clickListener);

    // 解除事件
    this.pageVisibility?.destory();
  }

  // eslint-disable-next-line @typescript-eslint/member-ordering
  stop() {
    this.clear();

    // 清空已存储的红包
    this.redpackItemList = {};
  }
}
export default RedpackRain;


jest.config.js
// https://doc.ebichu.cc/jest/docs/zh-Hans/configuration.html
module.exports = {
    collectCoverage: false,
    transform: {
        "^.+\\.tsx?$": "ts-jest",
    },
    testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
    // setupFiles: ["./__mocks__/localstorage.js", "./__mocks__/editableFn.js"],
    testPathIgnorePatterns: [
        "node_modules", // 默认
        "config",
        "esdocs",
        "docs",
    ],
    coverageDirectory: "coverage",
    // 'setupTestFrameworkScriptFile': '<rootDir>/__test__/setup.js',
    // testEnvironment: "<rootDir>/config/jsdom-env.js",
    testEnvironmentOptions: {
        userAgent:
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36",
        url: "https://example.org/",
        referrer: "https://example.com/",
    },
};


webpack.config.js
const path = require("path");
const webpack = require("webpack");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const TerserPlugin = require("terser-webpack-plugin");

const packageConfig = require("./package");

// https://webpack.docschina.org/configuration
module.exports = {
  mode: "production",
  entry: {
    index: "./src/index.ts",
  },
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "[name].js",
    libraryExport: "default",
    library: "RedpackRain",
    libraryTarget: "umd",
    globalObject: "this",
  },
  resolve: {
    extensions: [".tsx", ".ts", ".js"],
  },
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: {
          loader: "ts-loader",
        },
      },
    ],
  },
  optimization: {
    minimize: true,
    minimizer: [new TerserPlugin({ extractComments: false })],
  },
  plugins: [
    new CleanWebpackPlugin(),
    new webpack.BannerPlugin(
      `v${packageConfig.version} update: ${new Date().toLocaleString()}`
    ),
  ],
};


src/libs/redpackItem.ts
import {
  cancelAnimationFramePolyfill,
  requestAnimationFramePolyfill,
} from '../utils/animationFrame';
import { BubbleProps } from '../config/defaults';

interface RedpackItemProps {
  redpackId: number; // 红包标识
  redpackCtx: CanvasRenderingContext2D;
  bubbleCtx: CanvasRenderingContext2D;
  x: number; // 起始坐标x
  y: number; // 起始坐标y
  redpackImgUrl: string; // 图片地址
  width: number;
  height: number;
  speedMin: number;
  speedMax: number;
  bubble: BubbleProps;
  containerHeight: number;
  onDestoryed: (id: number) => void;
}

class RedpackItem {
  redpackId = -1; // 当前标识
  requestId = -1; // 动画的标识，用于清除动画
  x = 0;
  y = 0;
  speed = 1;
  redpackCtx: CanvasRenderingContext2D | null = null;
  bubbleCtx: CanvasRenderingContext2D | null = null;
  width = 0;
  height = 0;
  bubbleConfig: BubbleProps | null = null;
  redpackImgUrl = '';
  redpackImg: HTMLImageElement | null = null;
  containerHeight = 0;
  onDestoryed: ((id: number) => void) | null = null;
  angle = 0;
  ratio = 1;

  constructor({
    redpackId,
    redpackCtx,
    bubbleCtx,
    x,
    y,
    redpackImgUrl,
    width,
    height,
    speedMax,
    speedMin,
    bubble,
    containerHeight,
    onDestoryed,
  }: RedpackItemProps) {
    this.redpackId = redpackId;
    this.redpackCtx = redpackCtx;
    this.bubbleCtx = bubbleCtx;
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    this.speed = (speedMax - speedMin) * Math.random() + speedMin;
    this.redpackImgUrl = redpackImgUrl;
    this.containerHeight = containerHeight;
    this.onDestoryed = onDestoryed;
    this.bubbleConfig = bubble;

    const random = Math.random();
    const angle = ((random * 30 + 10) * Math.PI) / 180;
    this.angle = random < 0.5 ? angle : 0 - angle;
    this.ratio = random * 0.36 + 0.8;
  }

  async start() {
    this.redpackImg = await this.loadImage(this.redpackImgUrl);
    this.render();
  }

  // 加载图片
  loadImage(imgUrl: string): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
      const img = new Image();
      img.src = imgUrl;
      if (img.complete) {
        return resolve(img);
      }
      img.onload = () => {
        resolve(img);
      };
      img.onerror = reject;
    });
  }

  render() {
    const { redpackCtx } = this;
    if (!redpackCtx || !this.redpackImg) {
      return;
    }

    // 先清除当前位置
    // this.clear();
    if (this.y < this.containerHeight) {
      // 绘制下一个位置
      const nextx = this.x;
      const nexty = this.y + this.speed;

      redpackCtx.save();
      redpackCtx.beginPath();
      redpackCtx.rect(nextx, nexty, this.width, this.height);
      redpackCtx.translate(nextx + this.width / 2, nexty + this.height / 2);
      redpackCtx.scale(this.ratio, this.ratio);
      redpackCtx.rotate(this.angle);
      redpackCtx.drawImage(
        this.redpackImg,
        -this.width / 2,
        -this.height / 2,
        this.width,
        this.height,
      );
      redpackCtx.strokeStyle = 'transparent';
      redpackCtx.stroke();
      redpackCtx.restore();
      this.y = nexty;
    } else {
      // 超过边界，回调改元素被销毁的事件
      if (typeof this.onDestoryed === 'function') {
        this.onDestoryed(this.redpackId);
      }
    }
  }

  // 添加气泡
  async addBubble() {
    if (!this.bubbleCtx || !this.bubbleConfig) {
      return;
    }
    const { imgUrl, width, height, opacitySpeed, speed } = this.bubbleConfig;
    const bubbleImg = await this.loadImage(imgUrl);

    const nextx = this.x;
    let nexty = this.y;

    let alpha = 1;
    const bbs = () => {
      alpha -= opacitySpeed;
      if (!this.bubbleCtx) {
        return;
      }
      this.bubbleCtx.clearRect(nextx - 10, nexty - 10, width + 20, height + 20);

      if (alpha > 0) {
        this.bubbleCtx.save();
        this.bubbleCtx.globalAlpha = alpha;

        nexty -= speed;
        this.bubbleCtx.drawImage(bubbleImg, nextx, nexty, width, height);
        this.bubbleCtx.restore();
        requestAnimationFramePolyfill(bbs);
      }
    };

    requestAnimationFramePolyfill(bbs);
  }

  stop() {
    cancelAnimationFramePolyfill(this.requestId);
  }
}
export default RedpackItem;


src/config/defaults.ts
// 每个红包的配置
interface RedpackRainItem {
  speedMin: number;
  speedMax: number;
  imgUrl: string;
  width: number;
  height: number;
}

export interface BubbleProps {
  imgUrl: string;
  width: number;
  height: number;
  speed: number;
  opacitySpeed: number;
}

interface MonitorProps {
  fps: number;
}

interface RedpackRainProps {
  selector: HTMLElement;
  interval: number;
  eventType: 'click' | 'touchstart';
  redpack: RedpackRainItem;
  bubble: BubbleProps;
  onClick?: (isHit: boolean) => void;
  onMonitor?: (props: MonitorProps) => void;
}

const defaultsConfig: RedpackRainProps = {
  selector: document.body,
  interval: 1600,
  eventType: 'touchstart',
  redpack: {
    speedMin: 10,
    speedMax: 10,
    imgUrl: 'https://sola.gtimg.cn/aoi/sola/20201226100322_I1ltnkzJVc.png',
    width: 126,
    height: 174,
  },
  bubble: {
    imgUrl: 'https://sola.gtimg.cn/aoi/sola/20201225103914_2QQ9bXg2rU.png',
    width: 156,
    height: 111,
    speed: 5,
    opacitySpeed: 0.04,
  },
  onClick: () => {
    /* isHit */
  },
};
export default defaultsConfig;
